versiondb/memiavl/snapshot.go

```diff
@@ -0,0 +1,265 @@
+		// it use the minimal key from right subtree, but propagate the minimal key from left subtree.
+		nodeIndex, keyOffset, err := w.writeRecursive(node.Right())
+		if err != nil {
+			return 0, 0, err
+		}
+		binary.LittleEndian.PutUint64(buf[OffsetKey:], keyOffset)
+		binary.LittleEndian.PutUint32(buf[OffsetRight:], uint32(nodeIndex))
+
+		nodeIndex, minimalKeyOffset, err = w.writeRecursive(node.Left())
+		if err != nil {
+			return 0, 0, err
+		}
+		binary.LittleEndian.PutUint32(buf[OffsetLeft:], uint32(nodeIndex))
+	}
+
+	copy(buf[OffsetHash:OffsetHash+SizeHash], node.Hash())
+	if _, err := w.nodesWriter.Write(buf[:]); err != nil {
+		return 0, 0, err
+	}
+
+	i := w.nodesIndex
+	w.nodesIndex++
+	return i, minimalKeyOffset, nil
+}
+
+func loadSnapshot(snapshotDir string) (*PersistedBlobs, uint64, uint64, error) {
+	nodesFile := filepath.Join(snapshotDir, "nodes")
+	keysFile := filepath.Join(snapshotDir, "keys")
+	valuesFile := filepath.Join(snapshotDir, "values")
+	metadataFile := filepath.Join(snapshotDir, "metadata")
+
+	bz, err := os.ReadFile(metadataFile)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	version := binary.LittleEndian.Uint64(bz)
+	rootIndex := binary.LittleEndian.Uint64(bz[8:])
+
+	fpNodes, err := os.Open(nodesFile)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	nodes, nodesHandle, err := Mmap(fpNodes)
+	if err != nil {
+		fpNodes.Close()
+		return nil, 0, 0, err
+	}
+
+	if len(nodes) == 0 {
+		// empty tree
+		if rootIndex != 0 {
+			return nil, 0, 0, errors.New("corrupted snapshot, nodes are empty but rootIndex is not zero")
+		}
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		return nil, 0, 0, nil
+	}
+
+	if uint64(len(nodes)) != (rootIndex+1)*SizeNode {
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		return nil, 0, 0, fmt.Errorf("nodes file size %d don't match root node index %d", len(nodes), rootIndex)
+	}
+
+	fpKeys, err := os.Open(keysFile)
+	if err != nil {
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		return nil, 0, 0, err
+	}
+	keys, keysHandle, err := Mmap(fpKeys)
+	if err != nil {
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		_ = mmap.Munmap(keys, keysHandle)
+		fpKeys.Close()
+		return nil, 0, 0, err
+	}
+
+	fpValues, err := os.Open(valuesFile)
+	if err != nil {
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		_ = mmap.Munmap(keys, keysHandle)
+		fpKeys.Close()
+		return nil, 0, 0, err
+	}
+	values, valuesHandle, err := Mmap(fpValues)
+	if err != nil {
+		_ = mmap.Munmap(nodes, nodesHandle)
+		fpNodes.Close()
+		_ = mmap.Munmap(keys, keysHandle)
+		fpKeys.Close()
+		_ = mmap.Munmap(values, valuesHandle)
+		fpValues.Close()
+		return nil, 0, 0, err
+	}
+
+	blobs := &PersistedBlobs{
+		nodesFile:  fpNodes,
+		keysFile:   fpKeys,
+		valuesFile: fpValues,
+
+		nodes:  nodes,
+		keys:   keys,
+		values: values,
+
+		nodesHandle:  nodesHandle,
+		keysHandle:   keysHandle,
+		valuesHandle: valuesHandle,
+	}
+
+	return blobs, version, rootIndex, nil
```

Comment:
any concerns for not using `defer fpNodes.Close()`,  `defer fpKeys.Close()`, `defer fpValues.Close()`?